tools/src/main/scala/scala/scalanative/build/LLVM.scala

```diff
@@ -1,51 +1,89 @@
 package scala.scalanative
 package build
 
-import java.nio.file.{Files, Path, Paths}
+import java.nio.file.{Files, Path, Paths, StandardCopyOption}
 import java.util.Arrays
 import scala.collection.JavaConverters._
 import scala.util.Try
 import scala.sys.process._
 import scalanative.build.IO.RichPath
+import scalanative.build.Discover._
 
 /** Internal utilities to interact with LLVM command-line tools. */
 private[scalanative] object LLVM {
 
+  def unpackNativeCode(nativelib: NativeLib): Path =
+    if (NativeLib.isJar(nativelib)) unpackNativeJar(nativelib)
+    else copyNativeDir(nativelib)
+
   /**
-   * Unpack the `nativelib` to `workdir/lib`.
+   * Unpack the `nativelib` to `workdir/nativelib` where
+   * `nativelib` is the Scala Native lib or the name of
+   * a third party library that includes native code.
    *
    * If the same archive has already been unpacked to this location, this
    * call has no effects.
    *
    * @param nativelib The JAR to unpack.
    * @param workdir   The working directory. The nativelib will be unpacked
-   *                  to `workdir/lib`.
-   * @return The location where the nativelib has been unpacked, `workdir/lib`.
+   *                  to `workdir/nativelib`.
+   * @return The location where the nativelib has been unpacked, `workdir/nativelib`.
    */
-  def unpackNativelib(nativelib: Path, workdir: Path): Path = {
-    val lib         = workdir.resolve("lib")
-    val jarhash     = IO.sha1(nativelib)
-    val jarhashPath = lib.resolve("jarhash")
+  private def unpackNativeJar(nativelib: NativeLib): Path = {
+    val target      = nativelib.dest
+    val source      = nativelib.src
+    val jarhash     = IO.sha1(source)
+    val jarhashPath = target.resolve("jarhash")
     def unpacked =
-      Files.exists(lib) &&
+      Files.exists(target) &&
         Files.exists(jarhashPath) &&
         Arrays.equals(jarhash, Files.readAllBytes(jarhashPath))
 
     if (!unpacked) {
-      IO.deleteRecursive(lib)
-      IO.unzip(nativelib, lib)
+      IO.deleteRecursive(target)
+      IO.unzip(source, target)
       IO.write(jarhashPath, jarhash)
     }
 
-    lib
+    target
+  }
+
+  /**
+   * Copy project code from project `myproj` to `workdir/myproj`
+   * where is can be compiled and linked.
+   * Not used. Saved here in case non jar based is desired.
+   *
+   * @param nativelib    The native lib directory with native code.
+   * @param workdir   The working directory as described above.
+   * @return An Option with a Path to the `nativeCodeProject` work dir
+   *         or None if there is `NativeCodeProject is not configured.
```

Comment:
The `def` returns a `Path`, not an option. This comment seems obsolete.

Reply:
Moreover, this is *always* `nativelib.dest`, and it's the same in `unpackNativeJar`. It would be better to return `Unit` from both, and from `unpackNativeCode` as well, then directly use `nativelib.dest` at call site.

Reply:
This documentation has been fixed. The returned value from the function is used downstream. I need to look at this closer.

Reply:
I looked this closer and in `Build.build` the unpacking process returns the `NativeLib.dest` which is the generated destination path which is passed on later in the processing. From a functional programming point of view it is doing a side-effect and returning a value.